packages/marshal/src/marshal.js

```diff
@@ -124,49 +123,130 @@ export const makeMarshal = (
-        const errorId = nextErrorId();
+        const errorId = encodeRecur(nextErrorId());
+        assert.typeof(errorId, 'string');
         assert.note(err, X`Sent as ${errorId}`);
         marshalSaveError(err);
-        return harden({
-          [QCLASS]: 'error',
-          errorId,
-          message: `${err.message}`,
-          name: `${err.name}`,
-        });
+        return harden({ errorId, message, name });
       } else {
-        return harden({
-          [QCLASS]: 'error',
-          message: `${err.message}`,
-          name: `${err.name}`,
-        });
+        return harden({ message, name });
       }
     };
 
-    const encodeRemotableToCapData = val => {
-      const iface = getInterfaceOf(val);
-      // console.log(`serializeSlot: ${val}`);
-      return serializeSlot(val, iface);
-    };
+    if (serializeBodyFormat === 'capdata') {
+      /**
+       * @param {Passable} passable
+       * @param {InterfaceSpec} [iface]
+       * @returns {Encoding}
+       */
+      const encodeSlotToCapData = (passable, iface = undefined) => {
+        const { index, repeat } = encodeSlotCommon(passable);
 
-    const encode = makeEncodeToCapData({
-      encodeRemotableToCapData,
-      encodePromiseToCapData: serializeSlot,
-      encodeErrorToCapData,
-    });
+        if (repeat === true || iface === undefined) {
+          return harden({ [QCLASS]: 'slot', index });
+        } else {
+          return harden({ [QCLASS]: 'slot', iface, index });
+        }
+      };
 
-    const encoded = encode(root);
+      const encodeRemotableToCapData = (val, _encodeRecur) =>
+        encodeSlotToCapData(val, getInterfaceOf(val));
 
-    return harden({
-      body: JSON.stringify(encoded),
-      slots,
-    });
+      const encodePromiseToCapData = (promise, _encodeRecur) =>
+        encodeSlotToCapData(promise);
+
+      /**
+       * Even if an Error is not actually passable, we'd rather send
+       * it anyway because the diagnostic info carried by the error
+       * is more valuable than diagnosing why the error isn't
+       * passable. See comments in ErrorHelper.
+       *
+       * @param {Error} err
+       * @param {(p: Passable) => Encoding} encodeRecur
+       * @returns {Encoding}
+       */
+      const encodeErrorToCapData = (err, encodeRecur) => {
+        const errData = encodeErrorCommon(err, encodeRecur);
+        return harden({ [QCLASS]: 'error', ...errData });
+      };
+
+      const encodeToCapData = makeEncodeToCapData({
+        encodeRemotableToCapData,
+        encodePromiseToCapData,
+        encodeErrorToCapData,
+      });
+
+      const encoded = encodeToCapData(root);
+      const body = JSON.stringify(encoded);
+      return harden({
+        body,
+        slots,
+      });
+    } else if (serializeBodyFormat === 'smallcaps') {
+      /**
+       * @param {string} prefix
+       * @param {Passable} passable
+       * @param {string} [iface]
+       */
+      const serializeSlotToSmallcaps = (
```

Comment:
Consistency with `encodeSlotToCapData`:
```suggestion
      /**
       * @param {string} prefix
       * @param {Passable} passable
       * @param {InterfaceSpec} [iface]
       * @returns {string}
       */
      const serializeSlotToSmallcaps = (
```

Reply:
Done. Also renamed `serializeSlotToSmallcaps` to `encodeSlotToSmallcaps`